### ISSUE TITLE
Retrie is broken on Windows

### ISSUE THREAD

**Issue by bradrn** [haskell/haskell-language-server#2051]
Cross-posted from https://github.com/emacs-lsp/lsp-haskell/issues/128 on the recommendation of @michaelpj:

> When I try to apply a code action such as ‘Unfold’ or ‘Fold’, I get the following message:
> 
> ```
> LSP :: Please open an issue in lsp-mode for implementing `11160:retrie:retrieCommand'.
> 
> (error "Not enough arguments for format string")
> ```
> 

And the relevant part of the log from LSP:

> ```
> [Trace - 06:55:01 PM] Received response 'workspace/executeCommand - (95)' in 0ms.
> Result: {
>   "code": -32602,
>   "message": "error while parsing args for retrieCommand in plugin retrie: expected Bool, but encountered Null\narg = Object (fromList [(\"description\",String \"Unfold wSheet\"),(\"originatingFile\",String \"file:///c%3A/Users/bradn/Documents/Spreadsheet/supercell/src/Main.hs\"),(\"restrictToOriginatingFile\",Null),(\"rewrites\",Array [Object (fromList [(\"contents\",String \"Main.wSheet\"),(\"tag\",String \"Unfold\")])])])"
> }
> ```

**Comment by jneira** [comment]
Hi, thanks for the bug report, could you post the module where you are trying to trigger retry (or the relevant pragmas for retrie)?

**Comment by pepeiborra** [comment]
Could you also share a repro using VSCode?

**Comment by bradrn** [comment]
@jneira It fails for every module I’ve tried. And it’s not just retrie: all Wingman operations fail as well.

@pepeiborra I don’t use VSCode regularly, but I’ll try to reproduce it.

**Comment by bradrn** [comment]
@pepeiborra VSCode also gives me an error when I attempt to use Retrie or Wingman:

![image](https://user-images.githubusercontent.com/17613703/127490209-86ff7ab1-183b-41e7-a4ce-b3b0a85fee8a.png)

**Comment by pepeiborra** [comment]
Thanks @bradrn but we need more details to reproduce. Project, steps, versions of hls, ghc, etc.

**Comment by bradrn** [comment]
Project: this seems to happen for any project, even just plain Stack `simple` template.

GHC version: again, this seems to be independent of GHC version from what I can tell.

HLS version:

```
> .\haskell-language-server-wrapper.exe --version
haskell-language-server version: 1.2.0.0 (GHC: 8.10.4) (PATH: C:\Users\bradn\Documents\Haskell\hls\haskell-language-server-wrapper.exe) (GIT hash: be2071e985cb417f984ab00a1aad76dee02d6d0b)
```

**Comment by pepeiborra** [comment]
Can you attach a screen recording too please?

**Comment by bradrn** [comment]
https://user-images.githubusercontent.com/17613703/127492494-2d59799d-4fa7-4ab1-83d8-1da713d782e8.mp4

As shown in the video most other parts of HLS are working as expected.

**Comment by bradrn** [comment]
@pepeiborra Here’s a better recording:

https://user-images.githubusercontent.com/17613703/127493429-1c04d86c-43de-4692-8fca-8e70cf97bcf1.mp4

As shown towards the end of the video most parts of HLS are working correctly; it is only retrie and parts of Wingman which are buggy.

**Comment by jneira** [comment]
hmm from the path of the call to hls in the cli it seems you are on windows, no?

**Comment by bradrn** [comment]
Correct, I am on Windows 10.

**Comment by jneira** [comment]
ok, there is a existing issue about retrie and windows, although the symptoms are not the same afaics  #533

**Comment by bradrn** [comment]
Interesting… not sure if this is the same issue, but they certainly could be related. Do you know of any further tests I could try to diagnose this? (Though it’s getting late here and I don’t have time to do anything just right now.)

**Comment by jneira** [comment]
I just tried a retri `fold` and i am continue getting the same error as #533. Maybe your way to start emacs makes `xargs` to be in scope? Your error would be triggered after avoiding that

**Comment by bradrn** [comment]
Sorry, I’m not sure what you mean… are you saying that I should try *removing* `xargs` from PATH, or I should *add* it to PATH?

**Comment by jneira** [comment]
no, no, I think maybe you already have xargs in scope and the error you are getting is the next one
will try to reproduce yours

**Comment by jneira** [comment]
However wingman works correctly in my end

**Comment by isovector** [comment]
Logs for the wingman actions would be helpful as well. Thanks!

**Comment by bradrn** [comment]
@isovector Sorry, which logs do you want? Because as far as I can tell the logs are the same for Retrie and Wingman — both give similar parse errors.

**Comment by isovector** [comment]
@bradrn maybe just the whole log!

**Comment by jneira** [comment]
Mmm, one thing that share retrie and wingman is ghc-exactprint usage. No evidence to back the intuition though

**Comment by bradrn** [comment]
Curiouser and curiouser. I tried to test the various code actions available to confirm what does and doesn’t work, and to my surprise I could not find a single Wingman command which failed. I’m not sure why I reported that Wingman was failing, but whatever it was, it seems to be working perfectly now.

On the other hand, Retrie is still not working, but it seems to give different errors depending on which command I use:

- ‘Unfold’ and ‘Fold’ both give the parse error mentioned above
- ‘Unfold in current file’ gives no error, but has no effect either
- ‘Fold in current file’ fails with `ExitFailure 1`.

@isovector I tried collecting logs while testing, but I doubt you’ll be interested in all 70000+ lines!

**Comment by bradrn** [comment]
OK, now I’m thoroughly stumped. I decided to try Wingman in another project, and *this* time round it failed, with an absolutely bizarre error message:

```
LSP :: `workspace/executeCommand' with `1836:tactics:tacticsDestructLambdaCaseCommand' failed.

(error "C:\\sr\\ghc-8.10.4\\hyphenation-0.8.1-f8b6bb68cfad4b10e18172bc9e3eb5622f323a16\\share\\hyph-en-us.hyp.txt.gz: openBinaryFile: does not exist (No such file or directory)")
```

Any ideas?

**Comment by jneira** [comment]
Ugh, do you have the package installed in that location? `D:\sr` seems the dir stack sets as `%STACK_ROOT%`,  but `ghc-8.10.4\\hyphenation-0.8.1-f8b6bb68cfad4b10e18172bc9e3eb5622f323a16` is the typical path cabal uses to install packages with the v2-command. Maybe did you delete it?

**Comment by jneira** [comment]
> @isovector I tried collecting logs while testing, but I doubt you’ll be interested in all 70000+ lines! I’m not even sure where I would share a file of that size.

I think all lines emitted by wingman would have the plugin name, maybe extracting them from the log while reproducing the error (they should not be so much) would help

**Comment by bradrn** [comment]
> Ugh, do you have the package installed in that location? 

The path `C:\sr\ghc-8.10.4` doesn’t even exist on my machine. I’m not sure where it’s coming from, unless it’s hardcoded in the source (which it surely wouldn’t be‽).

> I think all lines emitted by wingman would have the plugin name, maybe extracting them from the log while reproducing the error (they should not be so much) would help

Good idea, I’ll try this.

**Comment by jneira** [comment]
> The path C:\sr\ghc-8.10.4 doesn’t even exist on my machine. I’m not sure where it’s coming from, unless it’s hardcoded in the source (which it surely wouldn’t be‽).

very interesting, those paths could be hardcoded while assembling the executable (although they are not directly in the source coide as you suspect), hou did you got the hls executable? did the vscode extension download for you?

**Comment by bradrn** [comment]
As mentioned already I’m using Emacs; the executables are simply downloaded from the GitHub releases page.

**Comment by jneira** [comment]
> As mentioned already I’m using Emacs; the executables are simply downloaded from the GitHub releases page.

oh, right, vscode downloads the executable from the same location. To investigate.

**Comment by isovector** [comment]
You're running into https://github.com/haskell/haskell-language-server/issues/2015

**Comment by bradrn** [comment]
Thanks @isovector!

**Comment by Tarmean** [comment]
I've been meaning to look into retrie on windows. My current thinking is that this is an escaping issues when passing backslash-based parts as stdin to xargs

    Prelude System.Process> readCreateProcessWithExitCode (shell "xargs grep  -l 'Var'") "C:/Users/name/Projects/natLog/library/Types.hs"
    (ExitSuccess,"C:/Users/name/Projects/natLog/library/Types.hs\n","")
    Prelude System.Process> readCreateProcessWithExitCode (shell "xargs grep  -l 'Var'") "C:\\Users\\name\\Projects\\natLog\\library\\Types.hs"
    (ExitFailure 123,"","grep: C:UsersnameProjectsnatLoglibraryTypes.hs: No such file or directory\n")

This command is build when retrie tries to filter down a list of paths [here](https://github.com/facebookincubator/retrie/blob/543907ea028ce82455d2cbd63ebe72d34b7b4940/Retrie/Options.hs#L469).

This command eventually runs this function in System.Process so maybe [the comment is related](https://github.com/haskell/process/blob/master/System/Process/Windows.hsc#L356).

The quickest fix probably would be to change the substitute \ with / in the paths [in the plugin](https://github.com/haskell/haskell-language-server/blob/master/plugins/hls-retrie-plugin/src/Ide/Plugin/Retrie.hs#L411).

**Comment by jneira** [comment]
@Tarmean wow, thanks for the further investigation, the fix could be convert the windows paths to unix style ones and does not seem to be too complex, would you be interested in make a pr?

**Comment by Tarmean** [comment]
The path conversion was really easy. Unfortunately the current System.Process implementation (well, hPutStr) also seems to mangle newlines in the stdin channel of the created process on windows by turning '\n' into '\r\n'.

The stdin handle is created as 'intPtrToPtr (-1)' and I don't know enough about how this magic works to say how badly things would break if one were to call hSetNewlineMode.

An more palatable ~~hack~~ alternative might be to inline the first xargs invocation, i.e. turning

    readCreateProcessWithExitCode (shell "xargs grep -l 'Fizz' | xargs grep -l 'Buzz'") paths

into

    readCreateProcessWithExitCode (shell "grep -l 'Fizz' $paths | xargs grep -l 'Buzz'") ""


Either way I'm currently out of time so I will have to experiment more later.

**Comment by jneira** [comment]
> Unfortunately the current System.Process implementation (well, hPutStr) also seems to mangle newlines in the stdin channel of the created process on windows by turning '\n' into '\r\n'.

I had to add a function to try to avoid that:

https://github.com/haskell/haskell-language-server/blob/2fef04193cfe2ac08372a008a5b5cfb6275d5bea/plugins/hls-hlint-plugin/src/Ide/Plugin/Hlint.hs#L437-L441

**Comment by Tarmean** [comment]
Replacing the usage of getTargetFiles with [this version](https://gist.github.com/Tarmean/5eeb906ac76d837425e043196a791358#file-gettargetfiles-hs-L83) works on windows, the additions are `filterChain0` and `toUnixFilepath`.

This is a gist and not a pullrequest to retrie because I noticed a couple of oddities and want to double check they aren't my fault. I think the weirdness are existing bugs in retrie, though

- Identifiers with `'` aren't found because it generates searches like `grep -L 'P''`
- Operators are treated really weirdly. Might be because fixity isn't loaded?

Here is an example of weird operator behavior

    {-# RULES "stepSum" forall x y. sum  (x : y) = x + (sum y) #-}
    foo :: Int
    foo = sum (1 : 2 : 3 : [])

    foo :: Int
    foo = 1 : 2 : 3 + (sum [])

**Comment by jneira** [comment]
I think this was closed by #2347, @Tarmean thanks!
